visualization/app/codeCharta/util/treeMapGenerator.ts

```diff
@@ -1,44 +1,108 @@
 import { hierarchy, HierarchyNode, HierarchyRectangularNode, treemap, TreemapLayout } from "d3"
 import { TreeMapHelper } from "./treeMapHelper"
-import { CodeMapHelper } from "./codeMapHelper"
 import { CodeMapNode, Node, NodeMetricData, State } from "../codeCharta.model"
 
+export type SquarifiedTreeMap = { treeMap: HierarchyRectangularNode<CodeMapNode>; height: number; width: number }
+
 export class TreeMapGenerator {
 	private static PADDING_SCALING_FACTOR = 0.4
 
 	public static createTreemapNodes(map: CodeMapNode, s: State, metricData: NodeMetricData[], isDeltaState: boolean): Node[] {
-		const squarifiedTreeMap: HierarchyRectangularNode<CodeMapNode> = this.getSquarifiedTreeMap(map, s)
 		const maxHeight = metricData.find(x => x.name == s.dynamicSettings.heightMetric).maxValue
 		const heightScale = (s.treeMap.mapSize * 2) / maxHeight
-		const nodesAsArray: HierarchyRectangularNode<CodeMapNode>[] = this.getNodesAsArray(squarifiedTreeMap)
-		return nodesAsArray.map(squarifiedNode => {
-			return TreeMapHelper.buildNodeFrom(squarifiedNode, heightScale, maxHeight, s, isDeltaState)
-		})
+
+		if (this.hasFixedFolders(map)) {
+			return this.buildSquarifiedTreeMapsForFixedFolders(map, s, heightScale, maxHeight, isDeltaState)
+		} else {
+			const squarifiedTreeMap = this.getSquarifiedTreeMap(map, s)
+			return squarifiedTreeMap.treeMap
+				.descendants()
+				.map(squarifiedNode => TreeMapHelper.buildNodeFrom(squarifiedNode, heightScale, maxHeight, s, isDeltaState))
+		}
 	}
 
-	private static getSquarifiedTreeMap(map: CodeMapNode, s: State): HierarchyRectangularNode<CodeMapNode> {
+	private static hasFixedFolders(map: CodeMapNode): boolean {
+		return !!map.children[0]?.fixed
+	}
+
+	private static buildSquarifiedTreeMapsForFixedFolders(
+		map: CodeMapNode,
+		state: State,
+		heightScale: number,
+		maxHeight: number,
+		isDeltaState: boolean
+	): Node[] {
 		const hierarchyNode: HierarchyNode<CodeMapNode> = hierarchy<CodeMapNode>(map)
-		const nodeLeafs: CodeMapNode[] = hierarchyNode.descendants().map(d => d.data)
-		const blacklisted: number = CodeMapHelper.numberOfBlacklistedNodes(nodeLeafs)
-		const nodesPerSide: number = 2 * Math.sqrt(hierarchyNode.descendants().length - blacklisted)
-		const mapLength: number = s.treeMap.mapSize * 2 + nodesPerSide * s.dynamicSettings.margin
-		const padding: number = s.dynamicSettings.margin * TreeMapGenerator.PADDING_SCALING_FACTOR
-		const treeMap: TreemapLayout<CodeMapNode> = treemap<CodeMapNode>()
-			.size([mapLength, mapLength])
-			.paddingOuter(padding)
-			.paddingInner(padding)
 
-		return treeMap(hierarchyNode.sum(node => this.calculateAreaValue(node, s)))
-	}
+		const nodes: Node[] = [TreeMapHelper.buildRootFolderForFixedFolders(map, heightScale, state, isDeltaState)]
+		const scale =
+			(state.treeMap.mapSize * 2 + this.getEstimatedNodesPerSide(hierarchyNode) * state.dynamicSettings.margin) / nodes[0].length
```

Comment:
I suggest to add comments to all of these calculations. Otherwise it's difficult to follow their logic.

Reply:
```ts
private static scaleAndTranslateSquarifiedNode(
		squarifiedNode: HierarchyRectangularNode<CodeMapNode>,
		fixedFolder: CodeMapNode,
		squarified: SquarifiedTreeMap,
		scale: number
	) {
		// Transform coordinates from local folder space to world space (between 0 and 100).
		const scaleX = fixedFolder.fixedPosition.width / squarified.width
		const scaleY = fixedFolder.fixedPosition.height / squarified.height

		// Scales to usual map-size of 500 matching the three-scene-size
		squarifiedNode.x0 = (squarifiedNode.x0 * scaleX + fixedFolder.fixedPosition.left) * scale
		squarifiedNode.x1 = (squarifiedNode.x1 * scaleX + fixedFolder.fixedPosition.left) * scale
		squarifiedNode.y0 = (squarifiedNode.y0 * scaleY + fixedFolder.fixedPosition.top) * scale
		squarifiedNode.y1 = (squarifiedNode.y1 * scaleY + fixedFolder.fixedPosition.top) * scale
	}
```

I added the comments on how the scaling process works (which was the most difficult part) here.
Basically, we translate and scale each node of a treeMap from a [0: INF] range, to a [0, 100] range. After this, each folder and buildings matches the `.cc.json` exactly. However, we have to scale the map again to match our mapSize, which we use in the threeScene. This was previously defined (not by me) as `(state.treeMap.mapSize * 2 + this.getEstimatedNodesPerSide(hierarchyNode) * state.dynamicSettings.margin)`.

Reply:
Thanks. I can't yet follow the comments but I hope they'll already help others :)

Please feel free to resolve this comment as it's not new code.